kymatio/scattering1d/convolution.py

```diff
@@ -0,0 +1,27 @@
+import torch.nn as nn
+import torch.nn.functional as F
+
+class Conv1dFFT(nn.Module):
+    def __init__(self, backend):
+        super(Conv1dFFT, self).__init__()
+        self.name = 'torch'
+        self.backend = backend
+
+    def preprocess_signal(self, x, padding=()):
+        if len(padding) != 0:
+            x = self.backend.pad(x, pad_left=padding[0], pad_right=padding[1])
+        x_hat = self.backend.rfft(x)
+        return x_hat
+
+    def convolution(self, x_hat, conv_filter, sampling_factor, domain='', ind_start=None, ind_end=None):
+        y_c = self.backend.cdgmm(x_hat, conv_filter)
+        y_hat = self.backend.subsample_fourier(y_c, 2**sampling_factor)
+        if domain == 'real':
```

Comment:
Ideally, this should be derived from `x` and `conv_filter`.

Reply:
explain?

Reply:
If `x` is real and `conv_filter` is Hermitian symmetric (that is, its IFT is real), the result of the convolution is real. Checking `x` is easy (we can do that in the `preprocess_signal` step). Checking the symmetry `conv_filter` is expensive, though.

The way to do it is to preprocess both the signal *and* the filter. The filter preprocessing would take place during the initialization of the scattering object (right after they're calculated). In the case of the FFT convolution, the only “preprocessing” would be to check the symmetry and store that as a flag. For other convolutions (like spatial or overlap–add), we would do other types of processing. The spatial convolution, for example, would compute the IFT and cut the result down so that only the non-zero elements are retained.

Longer-term, once we include padding into the convolution, this filter preprocessing would also include calculating the amount of padding that is needed. Short-term, we should try to include a preprocessing step for the filter as well (if only just to check the symmetry of the filter for now).

Reply:
I believe that `x` should always be real. I could add that, however the domain specified is asking "should we be taking the real part of this signal after computing the ifft." One way we could abstract this out would be having a operation that takes the real part of the ifft, so we call that in core whenever we want to get the real part of the tensor. This would be an operation in the general torch/numpy/tf backends. 

this is also what we were doing before R2C too.

Reply:
> I believe that x should always be real.

Right now, yes, but it's not necessarily always true. In the joint time–frequency transform, for example, you have to convolve complex signals sometimes.

> One way we could abstract this out would be having a operation that takes the real part of the ifft, so we call that in core whenever we want to get the real part of the tensor.

Right, but as you point out, that would basically undo a lot of the R2C stuff, which I think is useful. It would also prevent us from taking advantage of an efficient `irfft` if one ever materializes.

As I said, what we want to do is have another component that preprocesses the filters according to the convolution type we're doing. For the FFT, this just involves determining if the filter is real and computing the Fourier transform, storing both.